### PR TITLE
validate commit MVP implementation

### LINKED ISSUE #97: validate()
child of #89
child of #88

### LINKED ISSUE #171: stop manually maintaining the list of `HcApiFuncIndex` => invoke_*
the mapping of `HcApiFuncIndex` to `invoke_*` is entirely mechanical so we should try not to be maintaining it by hand

### LINKED ISSUE #176: mdbook for how to use/implement state reduce/actions

### PR DESCRIPTION
**Pull request by thedavidmeister**
fixes #97 
fixes #176 
fixes #171 

notes:

- it is not possible to dispatch a ValidateCommit action from inside a reducer for a Commit action, this causes the main action loop to hang, instead we do validate _then_ commit inside `invoke_commit`
- this doesn't touch on "incoming data from the DHT" as that doesn't exist yet, so no validate put exists yet
- this doesn't pull any data out of lifecycle functions through memory management, so we don't get messages for our fails in the WAT tests, only pass/fail/NI
- this is _NOT_ chasing the evolving spec app being developed by @lucksus 

changes:

- added action.rs to root
  - contains ActionWrapper, Action, Signal
- updated action/state docs in mdbook
- updated zome api functions implementation docs in mdbook
- updated lifecycle functions docs in mdbook
- updated "lifecycle of an entry" docs in mdbook
- started standardising reducer function signatures (more to do in followup)
- added validate_commit to invoke_commit from zome
  - invoke_commit will only dispatch the commit action if validate_commit returns `Pass` or `NotImplemented`
- removed some checks to history length in tests that caused genesis edge cases to hang indefinitely
- fix stack overflow in `HolochainError` trait implementation for `to_string()`
- `Action` is now global rather than specific to a state slice, e.g. zome can send actions that are reduced by either zome or agent
- `ActionResponse` is now reducer module specific, e.g. `agent::state::ActionResponse`
- include `unwrap_to!` macro to help get at enum wrapped values more easily
- abstractions for lifecycle vs. zome api functions
  - zome api functions are implemented by HC and called by zomes
  - implementing zome api functions is simpler, more centralised and less error prone
  - lifecycle functions are implemented by zomes and called by HC
  - genesis uses lifecycle function call
  - added `Defn` trait that ensures some consistency across lifecycle vs. API function definitions
- replaced adhoc index/mapping logic in zome API call dispatcher with a standardised interface into zome function definitions (to string, to index, to function, from string, etc.)
- initialization result logic extracted from genesis logic
- tweaks to stringification logic of various things
- ZomeFunctionResult has a dedicated reducer now
- added `ErrorLifecycleResult` as a zome WASM error code
- moar tests

todo/clarification in this tx:

- what is `Signal::ValidateEntry` and how does it relate to validate commit if at all? @ddd-mtl 
- reckon I could merge `ActionWrapper` and `Action`? @lucksus 
  - after pairing this morning, we decided we probably can as both have a unique id

followups:

- flatten action structure https://github.com/holochain/holochain-rust/issues/192
- trait for json round trip https://github.com/holochain/holochain-rust/issues/193
- standardise reduce function signatures https://github.com/holochain/holochain-rust/issues/194
- don't inspect history length directly https://github.com/holochain/holochain-rust/issues/195
- standard `ActionResponse` trait https://github.com/holochain/holochain-rust/issues/196
- flaky genesis test https://github.com/holochain/holochain-rust/issues/197
- standardise function call or action as key in nucleus state https://github.com/holochain/holochain-rust/issues/198
- unit test reducers better https://github.com/holochain/holochain-rust/issues/199
- lifecycle actions? https://github.com/holochain/holochain-rust/issues/200
- lifecycle function receive https://github.com/holochain/holochain-rust/issues/201
- as_fn in Defn trait https://github.com/holochain/holochain-rust/issues/202
- UUIDs in snowflakes for persister https://github.com/holochain/holochain-rust/issues/203
- document capabilities https://github.com/holochain/holochain-rust/issues/204
- get failure reasons back out of lifecycle functions https://github.com/holochain/holochain-rust/issues/205
- zome integration tests for validate commit and commit together https://github.com/holochain/holochain-rust/issues/206
- support nested action calls https://github.com/holochain/holochain-rust/issues/222
- better display handling for errors https://github.com/holochain/holochain-rust/issues/223

### CODE REVIEW

File: core/src/action.rs [hunk 1]
```diff
@@ -50,13 +50,17 @@ pub struct Action {
 }
 
 impl Action {
+    /// immutable constructor from &Signal
```
Comment (sphinxc0re) [inline core/src/action.rs]: What do you mean by `immutable constructor`?

Comment (thedavidmeister) [reply core/src/action.rs]: the action is immutable, i'll just say constructor

File: core/src/nucleus/ribosome/lifecycle/mod.rs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+};
+use num_traits::FromPrimitive;
+use std::{str::FromStr, sync::mpsc::Sender};
+
+// Lifecycle functions are zome logic called by HC actions
+// @TODO should each one be an action, e.g. Action::Genesis(Zome)?
+// @see https://github.com/holochain/holochain-rust/issues/200
+
+#[derive(FromPrimitive)]
+pub enum LifecycleFunction {
```
Comment (ddd-mtl) [inline core/src/nucleus/ribosome/lifecycle/mod.rs]: Confusion between `lifecycle` and `callback`
Rename to `callback` or list only the functions that belongs the the lifecycle capability

File: core/src/action.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    /// read only access to the internal Signal
+    pub fn signal(&self) -> Signal {
+        self.signal.clone()
+    }
+}
+
+impl Eq for Action {}
+
+#[derive(Clone, PartialEq, Hash, Debug)]
+pub enum Signal {
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/action.rs]: @ddd-mtl it has not been renamed signal, there were three (four?) structural levels before this branch, one was just implicit so i made it explicit and gave it a name

there's a followup to do what you're suggesting https://github.com/holochain/holochain-rust/issues/192

File: core/src/error.rs [hunk 4]
```diff
@@ -19,11 +19,18 @@ impl HolochainError {
...
+    /// @see https://github.com/holochain/holochain-rust/issues/193
+    pub fn to_json(&self) -> String {
+        format!("{{\"error\":\"{}\"}}", self.description())
+    }
 }
 
 impl fmt::Display for HolochainError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", self)
+        write!(f, "{:?}", self)
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/error.rs]: @ddd-mtl i don't know the subtleties

the high level is that what was there before blows a stack overflow on `to_string()` calls (which is also why i added that in a test below)... some implicit recursion somewhere

if you know the right fix we can drop it in, otherwise i'd be inclined to flag a followup for this

File: core/src/nucleus/ribosome/lifecycle/genesis.rs [hunk 5]
```diff
@@ -0,0 +1,91 @@
+use super::call;
+use action::ActionWrapper;
+use instance::Observer;
+use nucleus::ribosome::lifecycle::{
+    LifecycleFunction, LifecycleFunctionParams, LifecycleFunctionResult,
+};
+use std::sync::mpsc::Sender;
+
+pub fn genesis(
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/lifecycle/genesis.rs]: mmm, it's an abstraction that has little to do with the size of the contents of this function. i wanted to make a few things clear across all these functions:

- it locks the correct params enum to the function enum
- establishes a pattern to follow for the future as we add more implementations
- establishes where these functions are the same and where they can be extended/modified
- it's possible that the matching between errors and return values in `call` will shift into these functions, so in that case the logic will expand
- disambiguation `genesis` is immediately obvious whereas `call` requires thought (e.g. there's another related `call` function)

if you know a better way to do partial application of functions in rust maybe we can swap that out?

File: core/src/nucleus/ribosome/lifecycle/mod.rs (hunk 2 above)
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/lifecycle/mod.rs]: @ddd-mtl as i'm sure you picked up, i don't know how capabilities work, i cargo culted that in... still waiting on a pair session with @zippy to go over that but maybe you can run me through it too

we've been going back and forward on names like "hook", "zome function", etc.

callback doesn't seem right to me either because they aren't being passed around at runtime, let's have a quick straw poll in our next meeting to see if we can pick something

File: core/src/nucleus/ribosome/lifecycle/mod.rs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    /// genesis() -> bool
+    Genesis,
+
+    /// validate_commit() -> bool
+    ValidateCommit,
+
+    /// Communication Capability
+
+    /// receive(from : String, message : String) -> String
+    Receive,
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/lifecycle/mod.rs]: kk

File: core/src/nucleus/ribosome/lifecycle/mod.rs [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+    /// Error index for unimplemented functions
+    MissingNo = 0,
+
+    /// LifeCycle Capability
+
+    /// genesis() -> bool
+    Genesis,
+
+    /// validate_commit() -> bool
+    ValidateCommit,
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/lifecycle/mod.rs]: kk

File: core/src/nucleus/ribosome/lifecycle/mod.rs [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    /// receive(from : String, message : String) -> String
+    Receive,
+}
+
+impl FromStr for LifecycleFunction {
+    type Err = &'static str;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "genesis" => Ok(LifecycleFunction::Genesis),
+            "validate_commit" => Ok(LifecycleFunction::Genesis),
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/lifecycle/mod.rs]: yar, nice catch

[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/lifecycle/mod.rs]: i dropped some tests in for this

File: core/src/nucleus/ribosome/lifecycle/mod.rs [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    fn from_index(i: usize) -> Self {
+        match FromPrimitive::from_usize(i) {
+            Some(v) => v,
+            None => LifecycleFunction::MissingNo,
+        }
+    }
+
+    fn capabilities(&self) -> ReservedCapabilityNames {
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/lifecycle/mod.rs]: kk

File: core/src/nucleus/ribosome/lifecycle/validate_commit.rs [hunk 10]
```diff
@@ -0,0 +1,91 @@
+use super::call;
+use action::ActionWrapper;
+use instance::Observer;
+use nucleus::ribosome::lifecycle::{
+    LifecycleFunction, LifecycleFunctionParams, LifecycleFunctionResult,
+};
+use std::sync::mpsc::Sender;
+
+pub fn validate_commit(
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/lifecycle/validate_commit.rs]: yes, same deal

File: core/src/action.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/action.rs]: the purpose of the changes i made other than naming things is consolidation. i'm working towards:

- a flatter structure overall (e.g. something like what you suggested) so less `match`ing
- a single snowflake ID per action dispatch
- a single standard representation for key/values for action/result pairs across all state slices
- a single enum (rather than struct) that defines which types go where

happy to rename `Signal` if its confusing, there aren't many good words in this space though

File: core/src/nucleus/ribosome/api/mod.rs [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+    }
+}
+
+//--------------------------------------------------------------------------------------------------
+// Wasm call
+//--------------------------------------------------------------------------------------------------
+
+/// Object holding data to pass around to invoked API functions
+#[derive(Clone)]
+pub struct FunctionRuntime {
```
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/nucleus/ribosome/api/mod.rs]: @ddd-mtl this was part of something i did that i ended up rolling back, i'll change it back

File: core/src/error.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/error.rs]: @sphinxc0re any ideas for a real fix here?

[the remark this answers is not in the document]
Comment (thedavidmeister) [reply core/src/error.rs]: i've opened a followup https://github.com/holochain/holochain-rust/issues/223

File: core/src/nucleus/ribosome/lifecycle/mod.rs (hunk 2 above)
Comment (ddd-mtl) [reply core/src/nucleus/ribosome/lifecycle/mod.rs]: I'm also not comfortable with `callback` and agree with what you said. I was just sticking with the names used in the Go version.

File: doc/holochain_101/src/zome/callback_functions.md [hunk 12]
```diff
@@ -1,25 +1,24 @@
...
 
 Contrast this to a zome API function that is implemented by Holochain and called
 by the zome.
 
-Lifecycle functions are passed some parameters and expected to return one of
-three possible `LifecycleFunctionParams` variants:
+Callback functions are passed some parameters and expected to return one of
+three possible `CallbackParams` variants:
 
 - `Pass`: The lifecycle function has executed/validated successfully
```
Comment (sphinxc0re) [inline doc/holochain_101/src/zome/callback_functions.md]: `lifecycle` -> `callback`

Comment (thedavidmeister) [reply doc/holochain_101/src/zome/callback_functions.md]: you can just edit this in github UI... @sphinxc0re

**Comment by Connoropolous** [comment]
This was a heavy lift! Nice work David!

File: core/src/instance.rs [hunk 13]
```diff
@@ -277,16 +279,70 @@ pub mod tests {
...
+            println!("Waiting for ReturnZomeFunctionResult from genesis");
+            sleep(Duration::from_millis(10))
+        }
+
+        while instance
+            .state()
+            .history
+            .iter()
+            .find(|aw| match aw.action() {
+                Action::ReturnZomeFunctionResult(_) => true,
```
Comment (timotree3) [inline core/src/instance.rs]: this should be `Action::ReturnZomeInitializationResult`

File: doc/holochain_101/src/lifecycle_of_an_entry.md [hunk 14]
```diff
@@ -1 +1,18 @@
...
+Commit to source chain
+```
+
+Entries must be committed to the local source chain before they can be broadcast
+to the DHT.
+
+Every entry must pass a ValidateCommit lifecycle function check before it can be
+committed.
+
+If ValidateCommit is not implemented for the zome committing the entry then this
```
[the remark this answers is not in the document]
Comment (lucksus) [reply doc/holochain_101/src/lifecycle_of_an_entry.md]: Yes, this is contradictory to what I said at some point.
A missing validation function should be treated as **not valid**.

What we saw (and you can confirm or invalidated, @Connoropolous) a lot with people starting to write Holochain apps with the Go prototype was: not really implementing the validation function and just returning true. But actually that is the most important part of every Holochain app. So: for the current state it is not a problem to have it this way, but we should change this before the next release.

File: doc/holochain_101/src/zome/callback_functions.md [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+### ValidateCommit
+
+Canonical name: `validate_commit`
+Parameters: The candidate entry to be committed
+
+Called internally by the `commit` Zome API function.
+
+- `Pass`: the entry will be committed
+- `NotImplemented`: the entry will be committed
```
Comment (lucksus) [inline doc/holochain_101/src/zome/callback_functions.md]: no, should not be committed.

File: doc/holochain_101/src/lifecycle_of_an_entry.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (thedavidmeister) [reply doc/holochain_101/src/lifecycle_of_an_entry.md]: https://github.com/holochain/holochain-rust/issues/235
